src/mesh/topology.rs

```diff
@@ -19,43 +12,74 @@ use super::{Face, Mesh};
+}
+
+pub fn calculate_vertex_to_face_topology_from_components(
+    faces: &[Face],
+    vertex_count: u32,
+) -> Vec<VertexToFaceRelation> {
+    let mut v2f: Vec<SmallVec<[u32; 8]>> = vec![SmallVec::new(); cast_usize(vertex_count)];
+
+    for (face_index, face) in faces.iter().enumerate() {
+        let face_index_u32 = cast_u32(face_index);
+
+        match face {
+            Face::Triangle(triangle_face) => {
+                let vertices = &triangle_face.vertices;
+
+                for from_vertex in &[vertices.0, vertices.1, vertices.2] {
+                    if !v2f[cast_usize(*from_vertex)].contains(&face_index_u32) {
+                        v2f[cast_usize(*from_vertex)].push(face_index_u32);
+                    }
+                }
+            }
+        }
+    }
+
+    v2f
+}
+
 /// A topology containing neighborhood relations between faces. Two
 /// faces are neighbors if and only if they they share an unoriented
 /// edge.
-pub type FaceToFaceTopology = HashMap<u32, SmallVec<[u32; MAX_INLINE_NEIGHBOR_COUNT]>>;
+pub type FaceToFaceRelation = SmallVec<[u32; MAX_INLINE_NEIGHBOR_COUNT]>;
 
 /// Computes face to face topology for mesh.
-pub fn compute_face_to_face_topology(mesh: &Mesh) -> FaceToFaceTopology {
-    let mut f2f: HashMap<u32, SmallVec<[u32; 8]>> = HashMap::new();
+pub fn compute_face_to_face_topology(
+    mesh: &Mesh,
+    v2f: &[VertexToFaceRelation],
+) -> Vec<FaceToFaceRelation> {
+    let mut f2f: Vec<SmallVec<[u32; 8]>> = vec![SmallVec::new(); mesh.faces().len()];
 
-    for (from_face_index, from_face) in mesh.faces().iter().enumerate() {
-        let from_face_index_u32 = cast_u32(from_face_index);
-        let mut neighbors = SmallVec::new();
-
-        match from_face {
+    for (face_index, face) in mesh.faces().iter().enumerate() {
+        match face {
             Face::Triangle(triangle_face) => {
-                let [e1, e2, e3] = triangle_face.to_unoriented_edges();
-                for (to_face_index, to_face) in mesh.faces().iter().enumerate() {
-                    let to_face_index_u32 = cast_u32(to_face_index);
-
-                    match to_face {
-                        Face::Triangle(triangle_face) => {
-                            let face_contains_edge = triangle_face.contains_unoriented_edge(e1)
-                                || triangle_face.contains_unoriented_edge(e2)
-                                || triangle_face.contains_unoriented_edge(e3);
-                            if face_contains_edge
-                                && from_face_index != to_face_index
-                                && !neighbors.contains(&to_face_index_u32)
-                            {
-                                neighbors.push(to_face_index_u32);
-                            }
-                        }
+                let vertices = &triangle_face.vertices;
+                for first_vertex_in_face in &v2f[cast_usize(vertices.0)] {
```

Comment:
Isn't this a face index? The current name suggests it is a vertex index. Hmm, maybe something like `face_for_first_vertex` would be more clear. Also below for the face for second vertex.

Also, I am not sure we want to rely on the optimizer here, so if possible, let's access the vectors only once and use those values below.

```rust
let faces_for_1st_vertex = &v2f[cast_usize(vertices.0)];
let faces_for_2nd_vertex = &v2f[cast_usize(vertices.1)];
let faces_for_3rd_vertex = &v2f[cast_usize(vertices.2)];
```